tests/integ/ml/test_transformer.py

```diff
@@ -0,0 +1,505 @@
+    )
+    onehotencoder = OneHotEncoder()
+    onehotencoder.input_cols = ["A", "B", "C"]
+    onehotencoder.output_cols = ["ONEHOTARRAY"]
+    model = onehotencoder.fit(df)
+    # test category table
+    df_table = session.table(model._category_table_name["A"])
+    Utils.check_answer(
+        df_table,
+        [Row(STATES_A=1, STATES_A_ENCODER=0), Row(STATES_A=2, STATES_A_ENCODER=1)],
+    )
+
+    res = model.transform(df, sparse=False)
+    Utils.check_answer(
+        res,
+        [
+            Row(A=1, B=4, C=5, ONEHOTARRAY="[\n  1,\n  0,\n  1,\n  0,\n  1,\n  0\n]"),
+            Row(A=2, B=3, C=6, ONEHOTARRAY="[\n  0,\n  1,\n  0,\n  1,\n  0,\n  1\n]"),
+            Row(A=2, B=4, C=5, ONEHOTARRAY="[\n  0,\n  1,\n  1,\n  0,\n  1,\n  0\n]"),
+        ],
+    )
+
+
+def test_input_cols_output_cols_not_match(session):
+    df = session.createDataFrame(
+        [[1, 4, 5], [2, 4, 5], [2, 3, 6]], schema=["a", "b", "c"]
+    )
+    onehotencoder = OneHotEncoder()
+    onehotencoder.input_cols = ["A", "B", "C"]
+    onehotencoder.output_cols = ["B", "C"]
+    model = onehotencoder.fit(df)
+    with pytest.raises(ValueError) as ex_info:
+        model.transform(df)
+    assert "The number of output column and input column does not match" == str(
+        ex_info.value
+    )
+
+
+def test_preset_category_not_match_real_category(session):
+    df = session.createDataFrame(
+        [[1, 4, 5], [2, 4, 5], [2, 3, 6]], schema=["a", "b", "c"]
+    )
+    onehotencoder = OneHotEncoder(category=[["2"], ["4", "3"], ["5", "6"]])
+    onehotencoder.input_cols = ["A", "B", "C"]
+    onehotencoder.output_cols = ["ONEHOTARRAY"]
+    model = onehotencoder.fit(df)
+    # test category table
+    df_table = session.table(model._category_table_name["A"])
+    Utils.check_answer(
+        df_table,
+        [Row(STATES_A="2", STATES_A_ENCODER=0)],
+    )
+    # merge and sparse is true
+    res = model.transform(df)
+    Utils.check_answer(
+        res,
+        [
+            Row(
+                A=1,
+                B=4,
+                C=5,
+                ONEHOTARRAY="[\n  5,\n  [\n    1,\n    3\n  ],\n  1\n]",
+            ),
+            Row(
+                A=2,
+                B=4,
+                C=5,
+                ONEHOTARRAY="[\n  5,\n  [\n    0,\n    1,\n    3\n  ],\n  1\n]",
+            ),
+            Row(
+                A=2,
+                B=3,
+                C=6,
+                ONEHOTARRAY="[\n  5,\n  [\n    0,\n    2,\n    4\n  ],\n  1\n]",
+            ),
+        ],
+    )
+    # merge and sparse is false
+    res = model.transform(df, sparse=False)
+    Utils.check_answer(
+        res,
+        [
+            Row(A=1, B=4, C=5, ONEHOTARRAY="[\n  0,\n  1,\n  0,\n  1,\n  0\n]"),
+            Row(A=2, B=3, C=6, ONEHOTARRAY="[\n  1,\n  0,\n  1,\n  0,\n  1\n]"),
+            Row(A=2, B=4, C=5, ONEHOTARRAY="[\n  1,\n  1,\n  0,\n  1,\n  0\n]"),
+        ],
+    )
+
+    # merge and sparse is true
```

Comment:
```suggestion
    # merge is false and sparse is true
```
Also fix the previous comment